docs/specs/tsd-008-partial-write-error-handling.md

```diff
@@ -0,0 +1,59 @@
+# Partial write error handling
+
+## Objective
+
+Provide a way to pass information about partial metric write errors from an
+output to the output model.
+
+## Keywords
+
+output plugins, write, error, output model, metric, buffer
+
+## Overview
+
+When output plugins serialize metric and/or send them to the service endpoint,
+single metrics might cause errors e.g. by not being serializable or by being
+rejected by the service on the server side.
+Currently, an output is only able to accept or reject the complete batch of
+metrics it receives from the output model. This causes issues if only a subset
+of metrics in the batch fails as the output has no way of telling the model
+(and in turn the output buffer) which metrics failed but can only accept or
+reject the whole batch. As a consequence, outputs need to "accept" the batch
+to avoid a requeueing of the batch for the next flush interval. This distorts
+statistics of accepted metrics and causes misleading log messages.
+Even worse, for outputs ending-up with partial writes, e.g. only the first half
+of the metrics can be written to the service, there is no way of only accepting
+the written metrics so they need to internally buffer the remaining ones.
+
+This specification aims at defining the handling of partially successful writes
+and introduces the concept of a special _write error_ type. That error type
+must reflect partial writes and partial serialization to overcome the
+aforementioned issues.
+
+To do so, the error must contain a list of successfully
+written metrics, which must be marked as __accepted__ and must be removed from
+the buffer. The error must contain a list of metrics fatally failed to be
```

Comment:
Should accepted metrics be guaranteed to be in the error? Or could they be inferred based on any errored metrics?

Reply:
I would appreciate your help with the formulation as mine is ambiguous and overly complex I feel. What I want to say is that we get the list of metrics that are accepted (aka. can be dropped from the buffer) as well as the list of metrics rejected e.g. due to serialization errors or similar (aka. can be dropped from the buffer). Now all metrics in the batch not belonging to one of the mentioned lists should be kept in the buffer and re-issued for writing with the next batch!

So either the error explicitly provides them (which we might want in the future) _or_ we need to infer the metrics to keep from not being in one of the other lists...

Reply:
I think relying on the error explicitly providing accepted metrics specifically is that its not something we can rely on since if all are accepted there would be no error. I think it makes the most sense to have the error only give information about which metrics had an error (retryable or otherwise) and if a metric isn't mentioned in the error it can be assumed as accepted. Does that sound reasonable?

Reply:
To be honest I would take another view: All metrics that should be **dropped** from the buffer should be mentioned in the error so we do have an implicit failback of "what's not mentioned should be kept" which is the safe spot IMO. That's what currently is done, all metrics "accepted" and all metrics "rejected" are in there, so everything else should be kept.

Reply:
I think that makes a lot of sense in a code flow perspective, I think my only hangup is with this just being called an "error," when the error describes metrics that are not errored but accepted properly (and in many cases, may only contain accepted metrics). Maybe it makes sense for this not to be an error but some other explicit return type that could contain an error field? But if you don't have an issue with this then I'm fine with it

Reply:
If you want to accept all metrics, simply return a `nil` error as described. The "error" should only be used if there are `metrics fatally failed`!